File: webapp/src/features/achievements/achievementIcons.ts
import { createElement } from 'react';
import type { LucideIcon } from 'lucide-react';
import {
  AlignLeft,
  Award,
  BookOpen,
  Camera,
  Code2,
  Github,
  Heart,
  Image,
  Layers,
  MessageSquare,
  Monitor,
  PenLine,
  Shield,
  Sparkles,
  Users,
} from 'lucide-react';
import type { AchievementCategory } from '@/contracts/achievementsApi';

/** Slug → icon for catalog badges (swap for custom assets later via same slot). */
export const ACHIEVEMENT_ICON_BY_SLUG: Record<string, LucideIcon> = {
  'everyone-gets-one': Heart,
  'daily-syntax-readit': Sparkles,
  'daily-syntax-reader': BookOpen,
  'daily-syntax-author': PenLine,
  'profile-picture': Camera,
  'background-banner': Image,
  'bio-complete': AlignLeft,
  'github-connected': Github,
  'stack-and-tools': Code2,
  'my-setup': Monitor,
  'follow-three-authors': Users,
  'follow-three-categories': Layers,
  'squad-up': Shield,
  'first-feedback': MessageSquare,
  'invite-one-friend': Users,
  'invite-five-friends': Users,
  'invite-ten-friends': Award,
};

export function resolveAchievementIcon(slug: string): LucideIcon {
  return ACHIEVEMENT_ICON_BY_SLUG[slug] ?? Award;
}

type AchievementSlugIconProps = Readonly<{
  slug: string;
  className?: string;
  strokeWidth?: number;
}>;

/** Renders a catalog icon without assigning a component variable during render. */
export function AchievementSlugIcon({ slug, className, strokeWidth }: AchievementSlugIconProps) {
  return createElement(resolveAchievementIcon(slug), { className, strokeWidth });
}

export const ACHIEVEMENT_CATEGORY_LABEL: Record<AchievementCategory, string> = {
  engagement: 'Engagement',
  profile: 'Profile',
  reading: 'Reading',
  social: 'Social',
  meta: 'Writing',
};

export const ACHIEVEMENT_CATEGORY_TILE: Record<
  AchievementCategory,
  { tile: string; badge: string }
> = {
  engagement: {
    tile: 'border-primary/45 bg-primary/10 text-primary',
    badge: 'border-primary/30 bg-primary/5 text-primary',
  },
  profile: {
    tile: 'border-violet-500/40 bg-violet-500/10 text-violet-700 dark:text-violet-300',
    badge: 'border-violet-500/25 bg-violet-500/5 text-violet-700 dark:text-violet-300',
  },
  reading: {
    tile: 'border-amber-500/45 bg-amber-500/10 text-amber-700 dark:text-amber-300',
    badge: 'border-amber-500/30 bg-amber-500/5 text-amber-700 dark:text-amber-300',
  },
  social: {
    tile: 'border-sky-500/45 bg-sky-500/10 text-sky-700 dark:text-sky-300',
    badge: 'border-sky-500/30 bg-sky-500/5 text-sky-700 dark:text-sky-300',
  },
  meta: {
    tile: 'border-emerald-500/45 bg-emerald-500/10 text-emerald-700 dark:text-emerald-300',
    badge: 'border-emerald-500/30 bg-emerald-500/5 text-emerald-700 dark:text-emerald-300',
  },
};

File: mobile/lib/widgets/achievements/achievement_icons.dart
import 'package:flutter/material.dart';

import '../../models/achievement_item.dart';

const achievementCategoryLabels = <AchievementCategory, String>{
  AchievementCategory.engagement: 'Engagement',
  AchievementCategory.profile: 'Profile',
  AchievementCategory.reading: 'Reading',
  AchievementCategory.social: 'Social',
  AchievementCategory.meta: 'Writing',
};

const _iconBySlug = <String, IconData>{
  'everyone-gets-one': Icons.favorite_outline_rounded,
  'daily-syntax-readit': Icons.auto_awesome_outlined,
  'daily-syntax-reader': Icons.menu_book_outlined,
  'daily-syntax-author': Icons.edit_outlined,
  'profile-picture': Icons.photo_camera_outlined,
  'background-banner': Icons.image_outlined,
  'bio-complete': Icons.format_align_left_rounded,
  'github-connected': Icons.code_rounded,
  'stack-and-tools': Icons.memory_outlined,
  'my-setup': Icons.desktop_windows_outlined,
  'follow-three-authors': Icons.people_outline_rounded,
  'follow-three-categories': Icons.layers_outlined,
  'squad-up': Icons.shield_outlined,
  'first-feedback': Icons.chat_bubble_outline_rounded,
  'invite-one-friend': Icons.person_add_outlined,
  'invite-five-friends': Icons.group_add_outlined,
  'invite-ten-friends': Icons.emoji_events_outlined,
};

IconData resolveAchievementIcon(String slug) {
  return _iconBySlug[slug] ?? Icons.emoji_events_outlined;
}

class AchievementCategoryStyle {
  const AchievementCategoryStyle({
    required this.tileBorder,
    required this.tileBackground,
    required this.tileForeground,
    required this.badgeBorder,
    required this.badgeBackground,
    required this.badgeForeground,
  });

  final Color tileBorder;
  final Color tileBackground;
  final Color tileForeground;
  final Color badgeBorder;
  final Color badgeBackground;
  final Color badgeForeground;
}

AchievementCategoryStyle achievementCategoryStyle(
  AchievementCategory category,
  Color primary,
) {
  switch (category) {
    case AchievementCategory.engagement:
      return AchievementCategoryStyle(
        tileBorder: primary.withValues(alpha: 0.45),
        tileBackground: primary.withValues(alpha: 0.1),
        tileForeground: primary,
        badgeBorder: primary.withValues(alpha: 0.3),
        badgeBackground: primary.withValues(alpha: 0.05),
        badgeForeground: primary,
      );
    case AchievementCategory.profile:
      return const AchievementCategoryStyle(
        tileBorder: Color(0x998B5CF6),
        tileBackground: Color(0x1A8B5CF6),
        tileForeground: Color(0xFF7C3AED),
        badgeBorder: Color(0x408B5CF6),
        badgeBackground: Color(0x0D8B5CF6),
        badgeForeground: Color(0xFF7C3AED),
      );
    case AchievementCategory.reading:
      return const AchievementCategoryStyle(
        tileBorder: Color(0x99F59E0B),
        tileBackground: Color(0x1AF59E0B),
        tileForeground: Color(0xFFD97706),
        badgeBorder: Color(0x40F59E0B),
        badgeBackground: Color(0x0DF59E0B),
        badgeForeground: Color(0xFFD97706),
      );
    case AchievementCategory.social:
      return const AchievementCategoryStyle(
        tileBorder: Color(0x990EA5E9),
        tileBackground: Color(0x1A0EA5E9),
        tileForeground: Color(0xFF0284C7),
        badgeBorder: Color(0x400EA5E9),
        badgeBackground: Color(0x0D0EA5E9),
        badgeForeground: Color(0xFF0284C7),
      );
    case AchievementCategory.meta:
      return const AchievementCategoryStyle(
        tileBorder: Color(0x9910B981),
        tileBackground: Color(0x1A10B981),
        tileForeground: Color(0xFF059669),
        badgeBorder: Color(0x4010B981),
        badgeBackground: Color(0x0D10B981),
        badgeForeground: Color(0xFF059669),
      );
  }
}
